Enable users to authenticate using GitHub enterprise

fragsalat opened the pull request:
With merging and releasing my PR on the socialauth plugin https://github.com/3pillarlabs/socialauth/pull/95 it is possible to connect to an GitHub Enterprise server. With these PR I enable the users to authenticate on there GoCD servers via GitHub Enterprise. To do that they have to specify the access_token_url, authorize_url and api_url in the plugin configuration. If they are leaved blank the default public github urls are used.

ketan commented on github/src/main/resources/plugin-settings.template.html:
```diff
@@ -29,6 +29,21 @@
+    <input type="text" ng-model="api_url" ng-required="true" placeholder="https://github.com/api/v3"/>
+    <span class="form_error" ng-show="GOINPUTNAME[api_url].$error.server">{{ GOINPUTNAME[api_url].$error.server }}</span>
+</div>
+<div class="form_item_block">
```
Since these fields are optional, and are only used on GH enterprise, could you mark the label with `required for GH enterprise` and change the placeholder text to use `https://github.companyname.com/...` instead of `github.com` to avoid confusion?

ketan commented on github/src/main/java/com/tw/go/plugin/provider/github/GitHubProvider.java:
```diff
@@ -131,11 +142,30 @@ private boolean isAMemberOfOrganization(GithubPluginSettings pluginSettings, Use
+        // Connect to either enterprise github server or public github
+        if (pluginSettings.getApiUrl().isEmpty()) {
+            if (pluginSettings.containsUsernameAndPassword()) {
+                LOGGER.debug("Create Github connection to public Github using username and password");
```
`Github` is supposed to be written as `GitHub`, please. Trademarks and all :)

ketan commented on github/src/main/java/com/tw/go/plugin/provider/github/GitHubProvider.java:
```diff
@@ -131,11 +142,30 @@ private boolean isAMemberOfOrganization(GithubPluginSettings pluginSettings, Use
+                LOGGER.debug("Create Github connection to enterprise Github with token");
+                github = GitHub.connectToEnterprise(pluginSettings.getApiUrl(), pluginSettings.getOauthToken());
+            }
+        }
```
This might be easier to read if the long method was cut into a few methods using a simple extract method refactoring —

``` java
if (pluginSettings.getApiUrl().isEmpty()) {
  return publicGithubConnection(pluginSettings);
} else {
  return enterpriseGithubConnection(pluginSettings);
}
```

fragsalat commented:
I will change this tomorrow. Thanks for your fast response😄

fragsalat commented:
@ketan I fixed your feedback a bit differently. Instead of adding something to the labels I added a checkbox to toggle enterprise mode. Only if enterprise mode is enabled those fields are visible with an star sign.

ketan commented:
> @ketan I fixed your feedback a bit differently. Instead of adding something to the labels I added a checkbox to toggle enterprise mode. Only if enterprise mode is enabled those fields are visible with an star sign.

Thanks that's even better! I'll give it a quick spin and merge this in a day or two.

fragsalat commented:
Thank you very much. Can you please mention me when you released a new version? Should be 2.2 I though.

ketan commented:
@fragsalat A new release with your change is out - https://github.com/gocd-contrib/gocd-oauth-login/releases/tag/v2.2

fragsalat commented:
Thank you men :)
